plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +243,179 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            params.put("apirelativepath", "/nuage/api/" + apiVersion);
+            params.put("retrycount", String.valueOf(MoreObjects.firstNonNull(cmd.getApiRetryCount(), clientDefaults.get("DEFAULT_API_RETRY_COUNT"))));
+            params.put("retryinterval", String.valueOf(MoreObjects.firstNonNull(cmd.getApiRetryInterval(), clientDefaults.get("DEFAULT_API_RETRY_INTERVAL"))));
+
+            final Map<String, Object> hostdetails = new HashMap<String, Object>();
+            hostdetails.putAll(params);
             resource.configure(cmd.getHostName(), hostdetails);
 
-            final Host host = _resourceMgr.addHost(zoneId, resource, Host.Type.L2Networking, params);
+            Host host = _resourceMgr.addHost(zoneId, resource, Host.Type.L2Networking, params);
             if (host != null) {
-                return Transaction.execute(new TransactionCallback<NuageVspDeviceVO>() {
-                    @Override
-                    public NuageVspDeviceVO doInTransaction(TransactionStatus status) {
-                        NuageVspDeviceVO nuageVspDevice = new NuageVspDeviceVO(host.getId(), physicalNetworkId, ntwkSvcProvider.getProviderName(), deviceName);
-                        _nuageVspDao.persist(nuageVspDevice);
+                NuageVspDeviceVO nuageVspDevice = new NuageVspDeviceVO(host.getId(), physicalNetworkId, ntwkSvcProvider.getProviderName(), deviceName);
+                _nuageVspDao.persist(nuageVspDevice);
+
+                DetailVO detail = new DetailVO(host.getId(), "nuagevspdeviceid", String.valueOf(nuageVspDevice.getId()));
+                _hostDetailsDao.persist(detail);
+
+                ConfigurationVO cmsIdConfig = _configDao.findByName("nuagevsp.cms.id");
+                host = findNuageVspHost(nuageVspDevice.getHostId());
+                SyncNuageVspCmsIdCommand syncCmd = new SyncNuageVspCmsIdCommand(SyncType.REGISTER, null);
+                SyncNuageVspCmsIdAnswer answer = (SyncNuageVspCmsIdAnswer) _agentMgr.easySend(nuageVspDevice.getHostId(), syncCmd);
+                if (answer != null && answer.getSuccess()) {
+                    registerNewNuageVspDevice(cmsIdConfig, nuageVspDevice.getId() + ":" + answer.getNuageVspCmsId());
+
+                    detail = new DetailVO(host.getId(), "nuagevspcmsid", answer.getNuageVspCmsId());
+                    _hostDetailsDao.persist(detail);
+                    hostdetails.put("nuagevspcmsid", answer.getNuageVspCmsId());
+                    resource.configure((String) hostdetails.get("hostname"), hostdetails);
+
+                    auditDomainsOnVsp((HostVO) host, true, false);
+                } else {
+                    throw new CloudRuntimeException("Failed to register CMS ID");
+                }
+                return nuageVspDevice;
+            } else {
+                throw new CloudRuntimeException("Failed to add Nuage Vsp Device due to internal error.");
+            }
+        } catch (Exception e) {
```

Comment:
Need to catch <code>Exception</code> because of <code>Map<String, Object> clientDefaults = clientLoader.getNuageVspManagerClient().getClientDefaults();</code>, which declares <code>throws Exception</code>